napari/layers/image/_async_image.py

```diff
@@ -0,0 +1,765 @@
+    interpolation : str
+        Interpolation mode used by vispy. Must be one of our supported
+        modes.
+    rendering : str
+        Rendering mode used by vispy. Must be one of our supported
+        modes.
+    iso_threshold : float
+        Threshold for isosurface.
+    attenuation : float
+        Attenuation rate for attenuated maximum intensity projection.
+
+    Extended Summary
+    ----------
+    _data_view : array (N, M), (N, M, 3), or (N, M, 4)
+        Image data for the currently viewed slice. Must be 2D image data, but
+        can be multidimensional for RGB or RGBA images if multidimensional is
+        `True`.
+    _colorbar : array
+        Colorbar for current colormap.
+    """
+
+    _colormaps = AVAILABLE_COLORMAPS
+
+    def __init__(
+        self,
+        data,
+        *,
+        rgb=None,
+        colormap='gray',
+        contrast_limits=None,
+        gamma=1,
+        interpolation='nearest',
+        rendering='mip',
+        iso_threshold=0.5,
+        attenuation=0.05,
+        name=None,
+        metadata=None,
+        scale=None,
+        translate=None,
+        opacity=1,
+        blending='translucent',
+        visible=True,
+        multiscale=None,
+    ):
+        if isinstance(data, types.GeneratorType):
+            data = list(data)
+
+        # Determine if data is a multiscale
+        if multiscale is None:
+            multiscale, data = guess_multiscale(data)
+
+        # Determine initial shape
+        if multiscale:
+            init_shape = data[0].shape
+        else:
+            init_shape = data.shape
+
+        LOGGER.info(
+            "Image.__init__ multiscale=%d init_shape=%s",
+            multiscale,
+            init_shape,
+        )
+
+        # Determine if rgb
+        if rgb is None:
+            rgb = guess_rgb(init_shape)
+
+        # Determine dimensionality of the data
+        if rgb:
+            ndim = len(init_shape) - 1
+        else:
+            ndim = len(init_shape)
+
+        super().__init__(
+            data,
+            ndim,
+            name=name,
+            metadata=metadata,
+            scale=scale,
+            translate=translate,
+            opacity=opacity,
+            blending=blending,
+            visible=visible,
+            multiscale=multiscale,
+        )
+
+        self.events.add(
+            interpolation=Event,
+            rendering=Event,
+            iso_threshold=Event,
+            attenuation=Event,
+        )
+
+        # Set data
+        self.rgb = rgb
+        self._data = data
+
+        self._slice = None
+
+        # TODO_ASYNC: creating self._slice here seems totally wrong. This
+        # echoes how we did things before, and it's currently needed because
+        # stuff in _init_ requires it.
+        #
+        # We should get rid of this somehow and not create the slice
+        # until _update_dims() is called at the end of __init__ which
+        # will call refresh() and _set_view_slice().
```

Comment:
👍

Reply:
This is totally gone with recent changes. I basically redid image.py from scratch and now there are much less changes. We now create the ImageSlice identically to how we already created it. Which is great, less change.